yarpl/include/yarpl/flowable/EmitterFlowable.h

```diff
@@ -26,25 +26,20 @@ class EmiterBase : public virtual Refcounted {
  * This is synchronous: the emit calls are triggered within the context
  * of a request(n) call.
  */
-template<typename T>
-class EmiterSubscription : private Subscription, private Subscriber<T> {
+template <typename T>
+class EmiterSubscription : public virtual Subscription,
+                           public virtual Subscriber<T> {
   constexpr static auto kCanceled = credits::kCanceled;
   constexpr static auto kNoFlowControl = credits::kNoFlowControl;
 
  public:
   EmiterSubscription(
-      Reference <EmiterBase<T>> emiter,
-      Reference <Subscriber<T>> subscriber)
-      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {
-    // We expect to be heap-allocated; until this subscription finishes
-    // (is canceled; completes; error's out), hold a reference so we are
-    // not deallocated (by the subscriber).
-    Refcounted::incRef(*this);
-    subscriber_->onSubscribe(get_ref<Subscription>(this));
-  }
+      Reference<EmiterBase<T>> emiter,
+      Reference<Subscriber<T>> subscriber)
+      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {}
 
-  virtual ~EmiterSubscription() {
-    subscriber_.reset();
+  void init() {
+    subscriber_->onSubscribe(get_ref<Subscription>(this));
```

Comment:
Refcounting breaks for still-to-be-determined reasons; it seems bumping the refcount of ourselves in the ctor leads to us destroying ourselves in the ctor when the reference created there is destroyed (even if someone else holds a reference to us by the time we reach the end of the ctor).